src/Toml/BiMap.hs

```diff
@@ -51,245 +55,270 @@ module Toml.BiMap
+import qualified Data.Set as S
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Text.Lazy as TL
+import qualified Data.Text.Lazy.Encoding as TL
+
 
 ----------------------------------------------------------------------------
 -- BiMap concepts and ideas
 ----------------------------------------------------------------------------
 
 {- | Partial bidirectional isomorphism. @BiMap a b@ contains two function:
 
-1. @a -> Maybe b@
-2. @b -> Maybe a@
+1. @a -> Either e b@
+2. @b -> Either e a@
 -}
-data BiMap a b = BiMap
-    { forward  :: a -> Maybe b
-    , backward :: b -> Maybe a
+data BiMap e a b = BiMap
+    { forward  :: a -> Either e b
+    , backward :: b -> Either e a
     }
 
-instance Cat.Category BiMap where
-    id :: BiMap a a
-    id = BiMap Just Just
+instance Cat.Category (BiMap e) where
+    id :: BiMap e a a
+    id = BiMap Right Right
 
-    (.) :: BiMap b c -> BiMap a b -> BiMap a c
+    (.) :: BiMap e b c -> BiMap e a b -> BiMap e a c
     bc . ab = BiMap
         { forward  =  forward ab >=>  forward bc
         , backward = backward bc >=> backward ab
         }
 
 -- | Inverts bidirectional mapping.
-invert :: BiMap a b -> BiMap b a
+invert :: BiMap e a b -> BiMap e b a
 invert (BiMap f g) = BiMap g f
 
 -- | Creates 'BiMap' from isomorphism.
-iso :: (a -> b) -> (b -> a) -> BiMap a b
-iso f g = BiMap (Just . f) (Just . g)
+iso :: (a -> b) -> (b -> a) -> BiMap e a b
+iso f g = BiMap (Right . f) (Right . g)
 
 -- | Creates 'BiMap' from prism-like pair of functions.
-prism :: (field -> object) -> (object -> Maybe field) -> BiMap object field
-prism review preview = BiMap preview (Just . review)
+prism :: (field -> object) -> (object -> Either error field) -> BiMap error object field
+prism review preview = BiMap preview (Right . review)
+
+----------------------------------------------------------------------------
+-- BiMap error types
+----------------------------------------------------------------------------
+
+-- | Abstraction on BiMap with error.
+type TomlBiMap = BiMap TomlBiMapError
+
+prettyBiMapError :: TomlBiMapError -> Text
+prettyBiMapError = \case
+    WrongConstructor expect actual ->
+        "Invalid type constrictor. Expected " <> expect <> " Actual constructor: " <> actual
+    WrongValue (MatchError expect actual) ->
+        "Invalid value constructor. Expected " <> expect <> " Actual value: " <> toTxt actual
+    ArbitraryError text  -> "Invalid value: " <> text
 
 ----------------------------------------------------------------------------
 -- General purpose bimaps
 ----------------------------------------------------------------------------
 
 -- | Bimap for 'Either' and its left type
-_Left :: BiMap (Either l r) l
-_Left = prism Left (either Just (const Nothing))
+_Left :: (Show l, Show r) => TomlBiMap (Either l r) l
+_Left = prism Left $ \case
+    Left l -> Right l
+    x -> Left $ WrongConstructor "Left" (toTxt x)
```

Comment:
Let's create and expose `wrongConstructor` helper function to allow users to write such matches easily.